docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -910,6 +947,65 @@ spec:
             claimName: user-123-volume
 ```
 
+### State Transition
+
+In this section, we're explaining the TrainJob state transition (`.status.conditions`).
+The basic TrainJob state machine is the below, but the TrainJob has actual Jobs state as well.
+This means that If the TrainJob specifies the TrainingRuntime, the TrainJob has JobSet conditions as well,
+but actual Jobs states are added prefix like `JobSetCompleted`.
+
+Especially, if we specify the TrainingRuntime as a runtime, the TrainJob terminal condition (`Failed` or `Completed`) is decided
+based on the JobSet terminal state (`status.terminalState`) instead of computing from JobSet conditions.
+
+```mermaid
+stateDiagram-v2
+    #CREATION
+    state created_choice <<choice>>
+    [*] --> created_choice: TrainJob is submitted.
+    created_choice --> Created=True: Succeeded to build and deploy Jobs.
+    created_choice --> Created=False: Failed to build and deploy Jobs.
+    Created=False --> Created=False: Wait for updated appropriate TrainJob.
+    Created=False --> Created=True: Succeeded to build and deploy Jobs.
+
+    #SUSPENSION
+    state suspended_choice <<choice>>
+    Created=True --> suspended_choice: Handle TrainJob suspension.
+    suspended_choice --> Suspended=True: TrainJob is suspended.
+    Suspended=True --> Suspended=True: Wait for unsuspending.
+    Suspended=True --> Suspended=False: TrainJob is unsuspended.
+    suspended_choice --> Suspended=False: TrainJob is not suspended.
+
+    #FAILURE
+    state terminal_choice <<choice>>
+    Suspended=False --> terminal_choice: Actual Jobs go to terminal phase.
+    terminal_choice --> Failed=True: Actual Jobs (e.g., JobSet) failed.
+    Failed=True --> [*]
+
+    #COMPLETION
+    terminal_choice --> Completed=True: Actual Jobs (e.g., JobSet) completed.
+    Completed=True --> [*]
+```
+
+Hence, when we specify the JobSet as a runtime, after the TrainJob completed,
+we can see the TrainJob pseudo state in the following:
+
+```yaml
+[...]
+status:
+  conditions:
+  - type: Created
```

Comment:
Would it be helpful to add transition time and also reason for failure within conditions? I believe it would be help debugging failures

Reply:
This is pseudo conditions. So, actually we support all fields supported in the metav1.Conditions as you can see there: https://github.com/kubeflow/training-operator/blob/66ebecd12e689ffccc5b050560bd370c1bcc125d/pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go#L261-L262